File: Java/src/ToyORB/messages/MethodCallMessage.java
package ToyORB.messages;

import ToyORB.NetUtils;
import com.sun.istack.internal.NotNull;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.util.*;

public final class MethodCallMessage extends Message {
	public static final MessageType MESSAGE_TYPE = MessageType.METHOD_CALL;

	private final String methodName;
	private final LinkedHashMap<String, DataValue> args;

	public MethodCallMessage(@NotNull String methodName, @NotNull LinkedHashMap<String, Object> args) {
		super(MESSAGE_TYPE);
		this.methodName = Objects.requireNonNull(methodName);
		this.args = new LinkedHashMap<>();
		for (Map.Entry<String, Object> arg : args.entrySet()) {
			this.args.put(Objects.requireNonNull(arg.getKey()), DataValue.fromObject(arg.getValue()));
		}
	}

	@SuppressWarnings("unused")
	private MethodCallMessage(String methodName, LinkedHashMap<String, DataValue> args, Object dummy) {
		super(MESSAGE_TYPE);
		this.methodName = methodName;
		this.args = args;
	}

	public String getMethodName() {
		return methodName;
	}

	public Map<String, DataValue> getArguments() {
		return Collections.unmodifiableMap(args);
	}

	@Override
	protected void onWriteInto(DataOutputStream stream) throws IOException {
		NetUtils.writeStringInto(stream, methodName);
		stream.writeShort(args.size());
		for (Map.Entry<String, DataValue> arg : args.entrySet()) {
			NetUtils.writeStringInto(stream, arg.getKey());
			arg.getValue().writeInto(stream);
		}
	}

	@Override
	public String toString() {
		StringBuilder repr = new StringBuilder(128);
		String sep = "";

		repr.append("MethodCallMessage{").append(methodName).append("(");
		for (Map.Entry<String, DataValue> arg : args.entrySet()) {
			repr.append(sep);
			repr.append(arg.getKey()).append(": ");
			repr.append("(").append(arg.getValue().getType()).append(") ");
			repr.append(arg.getValue().getValue());
			sep = ", ";
		}
		repr.append(")}");

		return repr.toString();
	}

	@SuppressWarnings("unused")
	protected static final MessageReader<MethodCallMessage> READER = new MessageReader<MethodCallMessage>(MESSAGE_TYPE) {
		@Override
		public MethodCallMessage readMessageFrom(DataInputStream stream) throws IOException {
			String methodName = NetUtils.readStringFrom(stream);
			short argc = stream.readShort();
			LinkedHashMap<String, DataValue> args = new LinkedHashMap<>();
			for (int i = 0; i < argc; ++i) {
				String argName = NetUtils.readStringFrom(stream);
				DataValue argValue = DataValue.readFrom(stream);
				args.put(argName, argValue);
			}

			return new MethodCallMessage(methodName, args, null);
		}
	};
}

File: .NET Core/ToyORB/Messages/MethodCallMessage.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace ToyORB.Messages
{
    public class MethodCallMessage : Message
    {
        public static readonly MessageType MESSAGE_TYPE = MessageType.MethodCall;

        public string MethodName { get; }
        public IReadOnlyList<Tuple<string, DataValue>> Arguments { get; }

        public MethodCallMessage(string methodName, IList<Tuple<string, object>> arguments) : base(MESSAGE_TYPE)
        {
            MethodName = methodName ?? throw new ArgumentNullException(nameof(methodName));
            Arguments = arguments.Select(nv => Tuple.Create(nv.Item1, DataValue.FromObject(nv.Item2))).ToList();
        }

        private MethodCallMessage(string methodName, IList<Tuple<string, DataValue>> arguments) : base(MESSAGE_TYPE)
        {
            MethodName = methodName ?? throw new ArgumentNullException(nameof(methodName));
            Arguments = new List<Tuple<string, DataValue>>(arguments).AsReadOnly();
        }

        public override void WriteInto(BinaryWriter stream)
        {
            base.WriteInto(stream);
            stream.WriteBigEndian(MethodName);
            stream.WriteBigEndian((short) Arguments.Count);
            foreach (var kv in Arguments)
            {
                stream.WriteBigEndian(kv.Item1);
                kv.Item2.WriteInto(stream);
            }
        }

        public override string ToString()
        {
            string args = string.Join(", ", Arguments.Select(kv => string.Format("{0}: ({1}) {2}", kv.Item1, kv.Item2.Type, kv.Item2.Value)));
            return $"MethodCallMessage{{{MethodName}({args})}}";
        }

        protected internal class MessageReader : MessageReader<MethodCallMessage>
        {
            public MessageReader() : base(MESSAGE_TYPE) { }

            public override MethodCallMessage ReadMessageFrom(BinaryReader stream)
            {
                string methodName = stream.ReadStringBigEndian();
                short argc = stream.ReadInt16BigEndian();
                var arguments = new List<Tuple<string, DataValue>>();
                for (int i = 0; i < argc; ++i)
                {
                    string argName = stream.ReadStringBigEndian();
                    DataValue argValue = DataValue.ReadFrom(stream);
                    arguments.Add(Tuple.Create(argName, argValue));
                }
                return new MethodCallMessage(methodName, arguments);
            }
        }

        protected internal static MessageReader<MethodCallMessage> Reader = new MessageReader();
    }
}
